server/storage/gdrive.go

```diff
@@ -45,65 +44,66 @@ func NewGDriveStorage(clientJSONFilepath string, localConfigPath string, basedir
-	config, err := google.ConfigFromJSON(b, drive.DriveScope, drive.DriveMetadataScope)
-	if err != nil {
-		return nil, err
-	}
+	var httpClient *http.Client
+	var AuthType string
+
+	if strings.Contains(string(b), `"type": "service_account"`) {
+		AuthType = "service"
 
-	httpClient := getGDriveClient(ctx, config, localConfigPath, logger)
+		logger.Println("GDrive: using Service Account credentials")
+		config, err := google.JWTConfigFromJSON(b, drive.DriveScope, drive.DriveMetadataScope)
+		if err != nil {
+			return nil, err
+		}
+		httpClient = config.Client(ctx)
+	} else {
+		AuthType = "oauth"
+
+		if localConfigPath == "" {
+			return nil, fmt.Errorf("gdrive-local-config-path not set")
+		}
+
+		logger.Println("GDrive: using OAuth2 credentials")
+		config, err := google.ConfigFromJSON(b, drive.DriveScope, drive.DriveMetadataScope)
+		if err != nil {
+			return nil, err
+		}
+		httpClient = getGDriveClientFromToken(ctx, config, localConfigPath, logger)
+	}
 
 	srv, err := drive.NewService(ctx, option.WithHTTPClient(httpClient))
 	if err != nil {
 		return nil, err
 	}
 
-	storage := &GDrive{service: srv, basedir: basedir, rootID: "", localConfigPath: localConfigPath, chunkSize: chunkSize, logger: logger}
-	err = storage.setupRoot()
+	storage := &GDrive{service: srv, rootID: basedir, localConfigPath: localConfigPath, authType: AuthType, chunkSize: chunkSize, logger: logger}
+	err = storage.checkRoot()
 	if err != nil {
 		return nil, err
 	}
 
 	return storage, nil
 }
 
-func (s *GDrive) setupRoot() error {
-	rootFileConfig := filepath.Join(s.localConfigPath, gDriveRootConfigFile)
-
-	rootID, err := ioutil.ReadFile(rootFileConfig)
-	if err != nil && !os.IsNotExist(err) {
-		return err
-	}
-
-	if string(rootID) != "" {
-		s.rootID = string(rootID)
-		return nil
-	}
-
-	dir := &drive.File{
-		Name:     s.basedir,
-		MimeType: gDriveDirectoryMimeType,
-	}
-
-	di, err := s.service.Files.Create(dir).Fields("id").Do()
-	if err != nil {
-		return err
-	}
-
-	s.rootID = di.Id
-	err = ioutil.WriteFile(rootFileConfig, []byte(s.rootID), os.FileMode(0600))
-	if err != nil {
-		return err
+func (s *GDrive) checkRoot() error {
+	if s.rootID == "root" {
+		switch s.authType {
+		case "service":
+			return fmt.Errorf("GDrive: Folder \"root\" is not available when using Service Account credentials")
+		case "oauth":
+			s.logger.Println("GDrive: Warning: Folder \"root\" is not recommended.")
+		}
 	}
-
-	return nil
+	_, err := s.service.Files.Get(s.rootID).SupportsAllDrives(true).Do()
+	return err
 }
 
 func (s *GDrive) hasChecksum(f *drive.File) bool {
 	return f.Md5Checksum != ""
 }
 
 func (s *GDrive) list(nextPageToken string, q string) (*drive.FileList, error) {
-	return s.service.Files.List().Fields("nextPageToken, files(id, name, mimeType)").Q(q).PageToken(nextPageToken).Do()
+	return s.service.Files.List().Fields("nextPageToken, files(id, name, mimeType)").Q(q).PageToken(nextPageToken).SupportsAllDrives(true).IncludeItemsFromAllDrives(true).Do()
```

Comment:
why here you need `IncludeItemsFromAllDrives()` and not somewhere else?

Reply:
IncludeItemsFromAllDrives() is only need to List()
(checked on gdrive go docs)